### PR TITLE
You can now post to InterSpect

### PR DESCRIPTION
**Pull request by joepav**
- created express server in "backend" that can accept requests posted to it
- updates data on front-end via websockets
- Donald is planning to add toggle functionality to start/stop server

### CODE REVIEW

**Review by Donaldblanc — APPROVED** [review]
Thanks for handling this.  We will do some refactoring to toggle the server on or off put so far looks great.

File: src/main/containers/Panels.jsx [hunk 1]
```diff
@@ -1,18 +1,24 @@
-import React, { useState } from 'react';
+import React, { useState, useContext } from 'react';
 import styled from 'styled-components';
+import io from 'socket.io-client';
 import SourcePanel from './SourcePanel.jsx';
 import MockupsPanel from './MockupsPanel.jsx';
 import DestinationPanel from './DestinationPanel.jsx';
 import DataCanvas from './DataCanvas.jsx';
+import { TestsContext } from '../testsContext';
 // import { smallData, testsData } from '../dummyData';
```
Comment (conor-jh) [inline src/main/containers/Panels.jsx]: Let’s get rid of this dummy data import

File: src/main/containers/DataCanvas.jsx [hunk 2]
```diff
@@ -13,7 +13,7 @@ const DataCanvas = (props) => {
     return (
       <EmptyState>
         <h2>Let’s get some data</h2>
-        <p>To get started, send a request to any of your microservices—or open a data mockup</p>
+        <p>To get started, send a request to any of your microservices <br/>or post http://localhost:3001/posturl</p>
```
Comment (conor-jh) [inline src/main/containers/DataCanvas.jsx:16]: During our UI cleanup, I’d love to give this empty state some more thought as a team—just putting as a reminder here.

File: public/main.js [hunk 3]
```diff
@@ -12,6 +16,11 @@ if (isDev) {
 let mainWindow;
 
 function createWindow() {
+  expressApp.use(bodyParser.urlencoded({ extended: true }));
+  expressApp.use(bodyParser.json());
+
+  server.listen(3001);
```
Comment (conor-jh) [inline public/main.js:22]: I’m curious to hear what @Donaldblanc was saying when he mentioned randomizing across open ports. My (perhaps unfounded) concern with choosing something like `3001` would be conflicts with other running processes.

File: public/main.js [hunk 4]
```diff
@@ -32,5 +40,19 @@ function createWindow() {
   mainWindow.on('closed', () => {
     mainWindow = null;
   });
+
+  const poster = io.on('connection', (socket) => {
+    console.log('a user connected');
```
Comment (conor-jh) [inline public/main.js:62]: Is this `console.log` just for our benefit? It’d be *really* cool if we could put it in the Electron app chrome somehow. Something like a green dot and a “Connected" label, e.g.:
Screen Shot 2019-06-24 at 3 09 36 PM

File: public/main.js [hunk 5]
```diff
@@ -32,5 +40,19 @@ function createWindow() {
...
+
+    expressApp.post('/posturl', (request, response) => {
+      socket.emit('post_received', request.body);
+      response.status(200);
```
Comment (conor-jh) [inline public/main.js]: Let‘s add some basic error handling here. Thinking we should reject any non-JSON or non-XML data

File: src/main/containers/Panels.jsx [hunk 6]
```diff
@@ -34,6 +40,15 @@ const Panels = () => {
       options={dataTreeOptions} />
   );
 
+  socket.on('post_received', (postedData) => {
+    setData(postedData);
+    setTests([{ payload: postedData, status: '' }]);
```
Comment (conor-jh) [inline src/main/containers/Panels.jsx]: This should include an empty name property as well

File: src/main/containers/Panels.jsx [hunk 7]
```diff
@@ -34,6 +40,15 @@ const Panels = () => {
...
+
+    // clear old performance metrics on new post
```
Comment (conor-jh) [inline src/main/containers/Panels.jsx:48]: Good thinking 💯

**Review by conor-jh — CHANGES_REQUESTED** [review]
@joepav this is a _great_ add! Thanks for pushing us to do this and for taking it on. Just a few inline comments (mostly questions to consider for our polishing stage) and we should be good to go.

File: public/main.js (hunk 3 above)
Comment (Donaldblanc) [reply public/main.js:22]: @conorsexton  on the next iteration when adding the ability to toggle the server on or off, we can have Node randomly choose and open port by placing a 0 as an argument in the server constructor.

File: public/main.js (hunk 4 above)
Comment (Donaldblanc) [reply public/main.js:62]: I like this Idea,  We could implement that if we have time.

**Review by conor-jh — APPROVED** [review]

**Pull request merged**